src/controllers/qasm_controller.hpp

```diff
@@ -839,35 +854,56 @@ void QasmController::run_single_shot(const Circuit &circ, State_t &state,
 template <class State_t, class Initstate_t>
 void QasmController::run_circuit_with_noise(const Circuit &circ,
                                             const Noise::NoiseModel &noise,
-                                            uint_t shots, State_t &state,
+                                            const json_t &config, uint_t shots, State_t &state,
                                             const Initstate_t &initial_state,
+                                            const Method method,
                                             ExperimentData &data,
                                             RngEngine &rng) const {
-  // Sample a new noise circuit and optimize for each shot
+  // TODO: add opset to State class
+  Operations::OpSet state_opset;
+  state_opset.optypes = state.allowed_ops();
+  state_opset.gates = state.allowed_gates();
+  state_opset.snapshots = state.allowed_snapshots(); 
+
+  // Initialize fusion transpile pass
+  auto fusion_pass = transpile_fusion(method, circ, config);
+  Noise::NoiseModel dummy_noise;
+
   while (shots-- > 0) {
     Circuit noise_circ = noise.sample_noise(circ, rng);
     noise_circ.shots = 1;
-    if (noise_circ.num_qubits > circuit_opt_noise_threshold_) {
-      Noise::NoiseModel dummy;
-      optimize_circuit(noise_circ, dummy, state, data);
-    }
+    fusion_pass.optimize_circuit(noise_circ, dummy_noise, state_opset, data);
     run_single_shot(noise_circ, state, initial_state, data, rng);
   }
 }
 
 template <class State_t, class Initstate_t>
 void QasmController::run_circuit_without_noise(const Circuit &circ,
-                                               uint_t shots, State_t &state,
+                                               const json_t &config, uint_t shots, State_t &state,
                                                const Initstate_t &initial_state,
                                                const Method method,
                                                ExperimentData &data,
                                                RngEngine &rng) const {
   // Optimize circuit for state type
   Circuit opt_circ = circ;
-  if (opt_circ.num_qubits > circuit_opt_ideal_threshold_) {
-    Noise::NoiseModel dummy;
-    optimize_circuit(opt_circ, dummy, state, data);
-  }
+
+  // TODO: add opset to State class
+  Operations::OpSet state_opset;
+  state_opset.optypes = state.allowed_ops();
+  state_opset.gates = state.allowed_gates();
+  state_opset.snapshots = state.allowed_snapshots();
+  // Dummy noise model for transpiler passes
+  Noise::NoiseModel dummy_noise;
+
+  // Apply delay measure transpilation pass
+  Transpile::DelayMeasure measure_pass;
+  measure_pass.set_config(config);
+  measure_pass.optimize_circuit(opt_circ, dummy_noise, state_opset, data);
+
+  // Apply fusion transpilation pass
+  auto fusion_pass = transpile_fusion(method, circ, config);
+  fusion_pass.optimize_circuit(opt_circ, dummy_noise, state_opset, data);
+
```

Comment:
This is just a comment. If the number of posible transpile optimizations grows I'd create an optimization composite where you add optimizations and just call optimize_circuit on that object.

Reply:
Thats what we originally had and this PR removes :)
The problem was you needed to call passes at different points in execution, and different passes if there was noise or no noise. 

Maybe we could add a separate class that can collect optimization passes rather than adding a single one to the controller like we had before, then you can build separate collections to run at different points in the controller.

Reply:
Yeah, that's what I meant. Create an optimizer composite with the same interface as an optimizer, except that you can add optimizers to it and you just call `optimize_circuit` on this object. But, as you say, you can create them at any point and have different ones at different points of execution, is not a list for the controller. Something like this:

```
OptimizerComposite optimizations;
optimizations.add(optimizer1)
optimizations.add(optimizer2)
...
optimizations.optimize_circuit(opt_circ, dummy_noise, state_opset, data);
```
At this point, there is only one or two optimizers in this part of the code, so its ok. Just something to have in mind.

Reply:
Add it to the todo list!